api/middleware/logger.go

```diff
@@ -0,0 +1,137 @@
+package middleware
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	log "github.com/sirupsen/logrus"
+	jaeger "github.com/uber/jaeger-client-go"
+	macaron "gopkg.in/macaron.v1"
+)
+
+var (
+	logHeaders = false
+)
+
+type LoggingResponseWriter struct {
+	macaron.ResponseWriter
+	errBody []byte // the body in case it is an error
+}
+
+func (rw *LoggingResponseWriter) Write(b []byte) (int, error) {
+	if rw.ResponseWriter.Status() >= 400 {
+		rw.errBody = make([]byte, len(b))
+		copy(rw.errBody, b)
+	}
+	return rw.ResponseWriter.Write(b)
+}
+
+func Logger() macaron.Handler {
+	return func(ctx *Context) {
+		start := time.Now()
+		ctx.Resp = &LoggingResponseWriter{
+			ResponseWriter: ctx.Resp,
+		}
+		rw := ctx.Resp.(*LoggingResponseWriter)
+		ctx.MapTo(ctx.Resp, (*http.ResponseWriter)(nil))
+		ctx.Next()
+
+		// Only log:
+		// - requests that resulted in errors
+		// - requests on /render path
+		if rw.Status() >= 200 && rw.Status() < 300 && !strings.HasPrefix(ctx.Req.URL.Path, "/render") {
+			return
+		}
+
+		content := fmt.Sprintf("ts=%s", time.Now().Format(time.RFC3339Nano))
+
+		traceID, _ := extractTraceID(ctx.Req.Context())
+		if traceID != "" {
+			content += fmt.Sprintf(" traceID=%s", traceID)
+		}
+
+		err := ctx.Req.ParseForm()
+		if err != nil {
+			log.Errorf("Could not parse http request: %v", err)
+		}
+		paramsAsString := ""
+		if len(ctx.Req.Form) > 0 {
+			paramsAsString += "?"
+			paramsAsString += ctx.Req.Form.Encode()
+		}
+		content += fmt.Sprintf(" msg=\"%s %s%s (%v) %v\" orgID=%d", ctx.Req.Method, ctx.Req.URL.Path, paramsAsString, rw.Status(), time.Since(start), ctx.OrgId)
```

Comment:
I feel that the statuscode and request duration should be named fields in the log, rather then being put in 'msg'.  Also, is I feel that 'request' would be a better field name than 'msg'

Reply:
I agree. I did it that way to copy Cortex's way.

Reply:
I'd like to preserve that compatibility with Cortex if that's ok with you?